src/beast/beast/http/impl/write.ipp

```diff
@@ -247,79 +240,154 @@ public:
     }
 };
 
-template<class Stream, class Message, class Handler>
+template<class Stream, class Handler,
+    bool isRequest, class Body, class Headers>
 void
-write_op<Stream, Message, Handler, false>::
-    operator()(boost::system::error_code ec, std::size_t bytes_transferred)
+write_op<Stream, Handler, isRequest, Body, Headers>::
+operator()(error_code ec, std::size_t, bool again)
 {
     auto& d = *d_;
+    d.cont = d.cont || again;
     while(! ec && d.state != 99)
     {
         switch(d.state)
         {
         case 0:
+        {
+            d.wp.init(ec);
+            if(ec)
+            {
+                // call handler
+                d.state = 99;
+                d.s.get_io_service().post(bind_handler(
+                    std::move(*this), ec, 0, false));
+                return;
+            }
+            d.state = 1;
+            break;
+        }
+
         case 1:
         {
-            auto const result =
-                d.w.prepare(std::move(d.copy), ec);
+            auto const result = d.wp.w(std::move(d.copy), ec,
+                [&](auto const& buffers)
+                {
+                    // write headers and body
+                    if(d.wp.chunked)
+                        boost::asio::async_write(d.s,
+                            append_buffers(d.wp.sb.data(),
+                                chunk_encode(buffers)),
+                                    std::move(*this));
+                    else
+                        boost::asio::async_write(d.s,
+                            append_buffers(d.wp.sb.data(),
+                                buffers), std::move(*this));
+                });
             if(ec)
             {
                 // call handler
                 d.state = 99;
-                break;
+                d.s.get_io_service().post(bind_handler(
+                    std::move(*this), ec, false));
+                return;
             }
             if(boost::indeterminate(result))
             {
                 // suspend
-                d.state = 0;
                 d.copy = d.resume;
```

Comment:
I'm worried about complex ownership rules here. The writer takes ownership when it suspends (I think), it looks like `write_op` keeps a copy of `resume` and if the copy is moved from, re-initializes the copy from the saved `resume` here. Wouldn't it be simpler to let the writer make a copy if it needs one? I'm guessing there are other use cases I'm not thinking of, or I'm misreading the code.